src/awkward/contents/regulararray.py

```diff
@@ -856,6 +865,8 @@ def _combinations(self, n, replacement, recordlookup, parameters, axis, depth):
                 tocarry.append(ptr)
                 if self._backend.nplike.known_data:
                     tocarryraw[i] = ptr.ptr
+                else:
```

Comment:
> I don't think this line should be here - `ptr` is not referring to any external buffers.

You're right: I fixed all three cases in c9ba811f1c0cd47579c9ab3074360386f8ec9f1e.

> Also, shouldn't the various `Content._carry` be touching data?

It does, but it does so without special annotation because there's no `if nplike.known_data` hiding it from the type-tracer. For instance, in

https://github.com/scikit-hep/awkward/blob/b695e2fe78fbdb0a810a06a92e9fdd3aa3931d12/src/awkward/contents/listarray.py#L272-L283

the `self._starts[carry.data]` calls `__getitem__` on the `TypeTracerArray`, which finds its way into this case:

https://github.com/scikit-hep/awkward/blob/b695e2fe78fbdb0a810a06a92e9fdd3aa3931d12/src/awkward/_typetracer.py#L533-L540

which touches it.

The special `touch_data`/`touch_shape` annotations are only needed if the code that _would have called_ `TypeTracerArray`'s methods or `TypeTracer`'s functions doesn't get called because type-tracers are handled on a special code path. In this PR, I searched through all the places where `known_data` or `known_shape` are invoked and added annotations to the type-tracer code paths only.